orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/providers/gce/GoogleImageTagger.java

```diff
@@ -0,0 +1,192 @@
+            .build()
+          ).build()
+      );
+
+      originalTags.put(matchedImage.imageName, matchedImage.tags);
+    }
+
+    Map<String, Object> extraOutput = objectMapper.convertValue(stageData, Map.class);
+    extraOutput.put("targets", targetImages);
+    extraOutput.put("originalTags", originalTags);
+
+    return new ImageTagger.OperationContext(operations, extraOutput);
+  }
+
+  /**
+   * Return true iff the tags on the current machine image match the desired.
+   */
+  @Override
+  public boolean areImagesTagged(Collection<Image> targetImages, Stage stage) {
+    Collection<MatchedImage> matchedImages = findImages(
+      targetImages.stream().map(targetImage -> targetImage.imageName).collect(Collectors.toSet()),
+      stage
+    );
+
+    AtomicBoolean isUpserted = new AtomicBoolean(true);
+
+    for (Image targetImage : targetImages) {
+      MatchedImage matchedImage = matchedImages.stream()
+        .filter(m -> m.imageName.equals(targetImage.imageName))
+        .findFirst()
+        .orElse(null);
+
+      if (matchedImage == null) {
+        isUpserted.set(false);
+      } else {
+        targetImage.tags.entrySet().forEach(entry -> {
+          // assert tag equality
+          isUpserted.set(isUpserted.get() && entry.getValue().equals(matchedImage.tags.get(entry.getKey())));
+        });
+      }
+    }
+
+    return isUpserted.get();
+  }
+
+  @Override
+  public String getCloudProvider() {
+    return "gce";
+  }
+
+  private Collection<MatchedImage> findImages(Collection<String> imageNames, Stage stage) {
+    if (imageNames == null || imageNames.isEmpty()) {
+      imageNames = new HashSet<>();
+
+      // attempt to find upstream images in the event that one was not explicitly provided
+      Collection<String> upstreamImageIds = GoogleImageTaggerSupport.upstreamImageIds(stage);
+      if (upstreamImageIds.isEmpty()) {
+        throw new IllegalStateException("Unable to determine source image(s)");
+      }
+
+      for (String upstreamImageId : upstreamImageIds) {
+        // attempt to lookup the equivalent image name (given the upstream amiId/imageId)
+        List<Map> allMatchedImages = oortService.findImage(getCloudProvider(), upstreamImageId, null, null, null);
+        if (allMatchedImages.isEmpty()) {
+          throw new ImageNotFound(format("No image found (imageId: %s)", upstreamImageId), true);
+        }
+
+        String upstreamImageName = (String) allMatchedImages.get(0).get("imageName");
+        imageNames.add(upstreamImageName);
+
+        log.info(format("Found upstream image '%s' (executionId: %s)", upstreamImageName, stage.getExecution().getId()));
+      }
+    }
+
+    return imageNames.stream()
+      .map(targetImageName -> {
+        List<Map> allMatchedImages = oortService.findImage(getCloudProvider(), targetImageName, null, null, null);
+        Map matchedImage = allMatchedImages.stream()
+          .filter(image -> image.get("imageName").equals(targetImageName))
```

Comment:
It feels like this is adding a lot of logic that will need to be redone if we ever decide to respect the GCP "project" namespacing for images.